### DISCUSSION TITLE
Multiscale simulations of electromagnetism on complex domains with PINNs

### DISCUSSION THREAD

**Discussion by MSengler** [master-csmi/csmi#33] [Project Portfolio]
## Project presentation
I did my internship at the IRMA. My tutors are Emmanuel Franck and Victor Michel Dansac. This project is in partnership with the Thales group, a company specialized in electronics.

The objective of my internship is to obtain a minimal error on the prediction given by the PINNs of some problems of electromagnetism. These problems were given by Thales.  PINN is the acronym for Physics Informed Neural Network. It is used to resolve partial differential equations using a neural network. For this, we use the different residue as a loss function. The domains where we resolve our PDEs contain subdomains and consequently we have complex interface conditions between the two and the training of our neural network becomes more difficult. To implement this, we use Pytorch and the framework scimba.

During the internship, I have worked on a recent method based on Multi-scale algorithms which consist of taking several neural networks one after the other in order to reduce the error between the predicted solution and the exact one much more.

We do several training sessions to try finding the best one.  We try different configurations with different number of collocations points, learning rate, neural network architecture, weights of the different losses and others.  

## Results
I will present some of the problems which I have worked on and comment on each one.

![plotElectroGreedy](https://github.com/user-attachments/assets/dd09a90b-75df-4d46-a654-990f9788c866)

The first problem is a basic one from electrostatics. The method PINN with greedy gives really good results. We reach a maximum error of $10^{-6}$ and MSE of $10^{-8}$ on average 10 minutes.

![plotBimatElectroGreedy](https://github.com/user-attachments/assets/8c0f35ea-252c-4af7-b311-b05f8819d0cb)

We add a subdomain on our previous problem with a different permeability leading to some discontinuity. We need to impose interface conditions, but this one is badly defined on the corners of the rectangle. Consequently, the error remains important around the corner. For this one the maximum error is only $10^{-2}$. Nevertheless the MSE is better $10^{-6}$ on average. The time of the training is longer : some hours.

![plotMC](https://github.com/user-attachments/assets/2595d73e-2b10-4ce9-a582-d0df704e60ab)
![plotBC](https://github.com/user-attachments/assets/c4946c5f-a2f6-42f2-b153-585dab5a02ea)

We also worked on a magnetostatic problem, where we have a magnet in a vacuum. We have two problems : one with a rectangular and one circular magnet. We study at the same time the potential of the field and these derivatives. The two give quite similar measures. We reach a maximum error of  $10^{-3}$ and a MSE of $10^{-6}$. The error is principally localized on the top and bottom border of our magnet. The maximal error, and the errors on the derivatives are better on the circle due to the absence of the corners.

The multi-scale algorithms have improved our measures of error compared with the normal one. We have some factor of x7 on MSE and MAE, and much more for the first one problem. But, these methods also have difficulty in decreasing the maximum error on points where we have huge discontinuity or where the PDE is ill-defined (in the four corners of the rectangle).

## References
I put some of my references for people interested by the topic:

- M. Raissi, P. Perdikaris, and G.E. Karniadakis. Physics-informed neural net-
works: A deep learning framework for solving forward and inverse problems
involving nonlinear partial differential equations. Journal of Computational
Physics, 378:686707, 2019.

- Jonathan W. Siegel, Qingguo Hong, Xianlin Jin, Wenrui Hao, and Jinchao
Xu. Greedy training algorithms for neural networks and applications to
pdes. Journal of Computational Physics, 484:112084, 2023.

- Ziad Aldirany, Régis Cottereau, Marc Laforest, and Serge Prudhomme.
Multi-level neural networks for accurate solutions of boundary-value prob-
lems. Computer Methods in Applied Mechanics and Engineering, 419:116666,
2024
